README.md
# <img src="http://osmbuildings.org/logo.svg" width="100" height="88"/> OSM Buildings

OSM Buildings is a library for visualizing 3d building geometry.

<img src='http://i.imgur.com/pxh2Cpt.png' width='100%' />

- [Documentation](https://osmbuildings.org/documentation/)
- [Example](http://osmbuildings.org)

### Quick integration

Link all required libraries in your HTML head section. Files are provided in folder `/dist`.

````html
<head>
  <link href="https://cdn.osmbuildings.org/4.0.0/OSMBuildings.css" rel="stylesheet">
  <script src="https://cdn.osmbuildings.org/4.0.0/OSMBuildings.js"></script>
</head>

<body>
  <div id="map"></div>
````
In a script section initialize OSM Buildings and add it to a DOM container.

```` javascript
  var osmb = new OSMBuildings({
    container: 'map',
    position: { latitude: 52.52000, longitude: 13.41000 },
    zoom: 16,
    minZoom: 15,
    maxZoom: 22
  });

  osmb.addMapTiles(
    'https://{s}.tiles.mapbox.com/v3/osmbuildings.kbpalbpk/{z}/{x}/{y}.png',
    {
      attribution: '© Data <a href="http://openstreetmap.org/copyright/">OpenStreetMap</a> · © Map <a href="http://mapbox.com">Mapbox</a>'
    }
  );

  osmb.addGeoJSONTiles('http://{s}.data.osmbuildings.org/0.2/anonymous/tile/{z}/{x}/{y}.json');
````

### OSM Buildings server

There is also documentation of OSM Buildings Server side. See https://github.com/OSMBuildings/OSMBuildings/blob/master/docs/server.md


CHANGELOG.md

# 4.0.1 @ 2018-08-10 Bugfix

## Fixed

- fixed broken OBJ loading
- picking always returns object rather than undefined in case of miss
- compatibility with event property e.button restored for Safari
- navigator.hardwareConcurrency workaround for Safari


# 4.0.0 @ 2018-08-01

## Changed

- remove() function added to explicitly remove objects, markers or layers
- code optimizations for overall performance
- reduced loading lag by using web workers
- less CPU load by throttling renderer if there is no activity
- native markers enabled, source is SVG
- code base rewritten to ES6
- 'pointerup' event now returns target objects
- new 'load' event when viewer is ready
- multiple objects can be highlighted at once
- multiple objects can be hidden at once

- build process simplified
- sky texture dependency removed
- no picking on pointer movement removed
- some API methods and options changed
- loaded buildings are growing in again
- document events now acting on top rather than window

## Fixed

- Skillion roof with inner rings rendering fixed
- viewer resize fixed
- previous color adjustments reduced
- tile loading order reverted

## Breaking changes

- OSMBuildings() - options 'baseURL', 'showBackfaces', 'highlightColor', 'fogColor', 'effects[]' removed
- OSMBuildings now initializes with option 'container' instead of using method appendTo()
- methods appendTo(), show(), getTarget(), screenshot() removed
- events 'busy' and 'idle' removed
- setSize() - now takes width, height instead of { width: x, height: y }
- addGeoJSON() - option 'fadeIn' removed
- addOBJ() - option 'fadeIn' removed
- all keys/variables 'elevation' are renamed to 'altitude'


# 3.2.7 @ 2018-01-30

## Changed

- client side simplification removed
- improved roof creation
- sphere support added

## Fixed

- skillion roof direction
- touch event fix


# 3.2.4 @ 2017-09-03

## Changed

- rotation is now controlled by mouse button 2
- global min/max zoom enabled

## Fixed

- mouse event position
- better color validation
- low buildings were too dark
- better fallback for bad roof tagging
- destroy() was causing endless errors
- handling missing container height
- roof height calculation
- min/max zoom fixes


# 3.2.2 @ 2016-09-21

## Changed

- extended OSM roof shape support: Skillion, Gabled, Onion + Gabled as fallback for Hipped, Half-Hipped, Gambrel
- triangulation integrated from external module

## Fixed

- fadeIn option for GeoJSON tiles
- minZoom/maxZoom behavior
- touchmove positioning


# 3.1.0 @ 2016-07-27

## Changed

- protocol changes to https
- check for currently loading files removed - enables duplicates
- enabled 'loadfeature' event to act like 'modifier' option of objects (breaking compatibility)

## Fixed

- dynamic highlightColor restored for 'highlight(id, color)'


# 3.0.0 @ 2016-06-24

## Changed

- all OSM id's now have a prefix for their type to make them truly unique (breaking compatibility)
- GLMap and OSMBuildings are now integrated to one package (breaking compatibility)
- added 'fadeIn' option for adding GeoJSON or OBJ
- aligned pointer event positioning

## Fixed

- animation timing
- FOV calculation


# 2.4.3 @ 2016-05-09

## New

- Touch gesture for map tilt (2 finger vertical swipe)

## Changed

- internal GL tools reorganized
- Basemap now using native events


# 2.4.2 @ 2016-04-21

## Changed

Updates for CI automation


# 2.4.1 @ 2016-04-18

## New

- render effect 'outline'
- bower.json added
- option 'modifier' when adding GeoJSON or OBJ to change all properties upon load
- map events for move, zoom, rotate, tilt, rightclick, contextmenu
- OSM Buildings event for loadfeature

## Changed

- picking precision improved
- first steps in new documentation setup
- initial steps in CI
- geometry processing rewritten
- switch from color-coded depth calculation to depth textures
- don't render walls for building=roof
- code examples reorganized

## Fixed

- disable effects if GPU does not support them, fixes broken MSIE 11 on old machines
- pointerup event when releasing a finger
- map.getBounds() fixed, return format changed
- several destroy() methods fixed
- polyfill for Android gestures added
- pointer coordinates adjusted for maps with offset
- sun direction fixed
- several shadow Moire effects
- missing texture coordinates for OBJ


# 2.3.0 @ 2016-02-24

## New

- SSAO finished
- shadow casting on buildings and basemap
- sky added
- horizon fog added
- building windows added
- options to control effects and overall quality

## Changed

- zoom logic changed from scale-the-world to move-away

## Fixed

- OBJ files didn't work properly for picking
- CustomEvent polyfill had always been in effect
- potentially diverged animation frames fixed


# 2.2.1 @ 2015-11-18

## New

- timeline based animation support added
- buildings grow in on load
- new show() / hide() methods with custom selectors and optional duration
- map.setState() for setting position, zoom, rotation etc. all at once
- configuration option 'optimize' in order to switch anistrophic filtering and SSAO
- methods for 3d projection/unprojection: transforms a point on 2d screen into 3d space and vice versa
- screenshot() method

## Changed

- internal coordinate system now based on meters instead of pixel+zoom representation
- events using much less abstraction code
- JS files of GLMap and OSM Buildings are combined into one
- default data tile zoom set to 15
- whole rewrite of backend from PHP to NodeJS
- improved object picking performance
- distance based model for combining map tiles, increases performance
- data read operations optimized
- whole Color API rewritten
- better basemap tile blending on zoom level switch
- more optimization to detect visible tiles

## Fixed

- uniform and attribute locations on shader switch
- mouse events are now passed through, even if map is disabled
- roofHeight for unsupported shapes fixed
- populating buffers fixed for large OBJ files
- more destructor fixes
- restored idle/busy events

src/GLMap.js

var GLMap = function(containerId, options) {

  options = options || {};

  this._listeners = {};
  this._layers = [];
  this._container = document.getElementById(containerId);

  this.minZoom = parseFloat(options.minZoom) || 10;
  this.maxZoom = parseFloat(options.maxZoom) || 20;

  if (this.maxZoom < this.minZoom) {
    this.maxZoom = this.minZoom;
  }

  this._initState(options);
  this._initEvents(this._container);
  this._initRenderer(this._container);

  this.setDisabled(options.disabled);
};

GLMap.prototype = {

  _initState: function(options) {
    this._center = {};
    this._size = { width:0, height:0 };
    options = State.load(options);
    this.setCenter(options.center || { latitude:52.52000, longitude:13.41000 });
    this.setZoom(options.zoom || this.minZoom);
    this.setRotation(options.rotation || 0);
    this.setTilt(options.tilt || 0);

    this.on('change', function() {
      State.save(this);
    }.bind(this));

    State.save(this);
  },

  _initEvents: function(container) {
    this._startX = 0;
    this._startY = 0;
    this._startRotation = 0;
    this._startZoom = 0;

    this._hasTouch = ('ontouchstart' in window);
    this._dragStartEvent = this._hasTouch ? 'touchstart' : 'mousedown';
    this._dragMoveEvent  = this._hasTouch ? 'touchmove'  : 'mousemove';
    this._dragEndEvent   = this._hasTouch ? 'touchend'   : 'mouseup';

    addListener(container, this._dragStartEvent, this._onDragStart.bind(this));
    addListener(container, 'dblclick',   this._onDoubleClick.bind(this));
    addListener(document, this._dragMoveEvent, this._onDragMove.bind(this));
    addListener(document, this._dragEndEvent,  this._onDragEnd.bind(this));

    if (this._hasTouch) {
      addListener(container, 'gesturechange', this._onGestureChange.bind(this));
    } else {
      addListener(container, 'mousewheel',     this._onMouseWheel.bind(this));
      addListener(container, 'DOMMouseScroll', this._onMouseWheel.bind(this));
    }

    addListener(window, 'resize', this._onResize.bind(this));
  },

  _initRenderer: function(container) {
    var canvas = document.createElement('CANVAS');
    canvas.style.position = 'absolute';
    canvas.style.pointerEvents = 'none';

    container.appendChild(canvas);

    // TODO: handle context loss
    try {
      gl = canvas.getContext('experimental-webgl', {
        antialias: true,
        depth: true,
        premultipliedAlpha: false
      });
    } catch(ex) {
      throw ex;
    }

    addListener(canvas, 'webglcontextlost', function(e) {
      cancelEvent(e);
      clearInterval(this._loop);
    }.bind(this));

    addListener(canvas, 'webglcontextrestored', this._initGL.bind(this));

    this._initGL();
  },

  _initGL: function() {
    this.setSize({ width:this._container.offsetWidth, height:this._container.offsetHeight });

    gl.enable(gl.CULL_FACE);
    gl.enable(gl.DEPTH_TEST);

    this._loop = setInterval(this._render.bind(this), 17);
  },

  _render: function() {
    requestAnimationFrame(function() {
      gl.clearColor(0.5, 0.5, 0.5, 1);
      gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
//    for (var i = this._layers.length-1; i >= 0; i--) {
      for (var i = 0; i < this._layers.length; i++) {
        this._layers[i].render(this._projection);
      }
    }.bind(this));
  },

  _onDragStart: function(e) {
    if (this._isDisabled || (e.button !== undefined && e.button !== 0)) {
      return;
    }

    cancelEvent(e);

    if (e.touches !== undefined) {
      this._startRotation = this._rotation;
      this._startZoom = this._zoom;
      if (e.touches.length > 1) {
        return;
      }
      e = e.touches[0];
    }

    this._startX = e.clientX;
    this._startY = e.clientY;

    this._isDragging = true;
  },

  _onDragMove: function(e) {
    if (this._isDisabled || !this._isDragging) {
      return;
    }

    if (e.touches !== undefined) {
      if (e.touches.length > 1) {
        return;
      }
      e = e.touches[0];
    }

    var dx = e.clientX-this._startX;
    var dy = e.clientY-this._startY;
    var r = this._rotatePoint(dx, dy, this._rotation * Math.PI / 180);
    this.setCenter(unproject(this._origin.x-r.x, this._origin.y-r.y, this._worldSize));

    this._startX = e.clientX;
    this._startY = e.clientY;
  },

  _onDragEnd: function(e) {
    if (this._isDisabled || !this._isDragging) {
      return;
    }

    if (e.touches !== undefined) {
      if (e.touches.length > 1) {
        return;
      }
      e = e.touches[0];
    }

    this._isDragging = false;

    var dx = e.clientX-this._startX;
    var dy = e.clientY-this._startY;
    var r = this._rotatePoint(dx, dy, this._rotation * Math.PI / 180);
    this.setCenter(unproject(this._origin.x-r.x, this._origin.y-r.y, this._worldSize));
  },

  _rotatePoint: function(x, y, angle) {
    return {
      x: Math.cos(angle)*x - Math.sin(angle)*y,
      y: Math.sin(angle)*x + Math.cos(angle)*y
    };
  },

  _onGestureChange: function(e) {
    if (this._isDisabled) {
      return;
    }
    cancelEvent(e);
    this.setRotation(this._startRotation-e.rotation);
    this.setZoom(this._startZoom + (e.scale - 1));
  },

  _onDoubleClick: function(e) {
    if (this._isDisabled) {
      return;
    }
    cancelEvent(e);
    this.setZoom(this._zoom + 1, e);
  },

  _onMouseWheel: function(e) {
    if (this._isDisabled) {
      return;
    }
    cancelEvent(e);
    var delta = 0;
    if (e.wheelDeltaY) {
      delta = e.wheelDeltaY;
    } else if (e.wheelDelta) {
      delta = e.wheelDelta;
    } else if (e.detail) {
      delta = -e.detail;
    }

    var adjust = 0.2 * (delta > 0 ? 1 : delta < 0 ? -1 : 0);

    this.setZoom(this._zoom + adjust, e);
  },

  _onResize: function() {
    clearTimeout(this._resizeTimer);
    this._resizeTimer = setTimeout(function() {
      var container = this._container;
      if (this._size.width !== container.offsetWidth || this._size.height !== container.offsetHeight) {
        this.setSize({ width:container.offsetWidth, height:container.offsetHeight });
      }
    }.bind(this), 250);
  },

  _emit: function(type) {
    if (!this._listeners[type]) {
      return;
    }
    var listeners = this._listeners[type];
    for (var i = 0, il = listeners.length; i < il; i++) {
      listeners[i]();
    }
  },

  addLayer: function(layer) {
    this._layers.push(layer);
  },

  removeLayer: function(layer) {
    for (var i = 0; i < this._layers.length; i++) {
      if (this._layers[i] === layer) {
        this._layers[i].splice(i, 1);
        return;
      }
    }
  },

  setDisabled: function(flag) {
    this._isDisabled = !!flag;
  },

  on: function(type, listener) {
    var listeners = this._listeners[type] || (this._listeners[type] = []);
    listeners.push(listener);
    return this;
  },

  _setOrigin: function(origin) {
    this._origin = origin;
  },

  off: function(type, listener) {
    return this;
  },

  getZoom: function() {
    return this._zoom;
  },

  setZoom: function(zoom, e) {
    zoom = clamp(parseFloat(zoom), this.minZoom, this.maxZoom);

    if (this._zoom !== zoom) {
      if (!e) {
        this._zoom = zoom;
        this._worldSize = TILE_SIZE * Math.pow(2, zoom);
        this._setOrigin(project(this._center.latitude, this._center.longitude, this._worldSize));
      } else {
        var size = this.getSize();
        var dx = size.width /2 - e.clientX;
        var dy = size.height/2 - e.clientY;
        var geoPos = unproject(this._origin.x - dx, this._origin.y - dy, this._worldSize);

        this._zoom = zoom;
        this._worldSize = TILE_SIZE * Math.pow(2, zoom);

        var pxPos = project(geoPos.latitude, geoPos.longitude, this._worldSize);
        this._setOrigin({ x:pxPos.x+dx, y:pxPos.y+dy });
        this._center = unproject(this._origin.x, this._origin.y, this._worldSize);
      }

      this._emit('change');
    }

    return this;
  },

  getCenter: function() {
    return this._center;
  },

  setCenter: function(center) {
    center.latitude  = clamp(parseFloat(center.latitude),   -90,  90);
    center.longitude = clamp(parseFloat(center.longitude), -180, 180);

    if (this._center.latitude !== center.latitude || this._center.longitude !== center.longitude) {
      this._center = center;
      this._setOrigin(project(center.latitude, center.longitude, this._worldSize));
      this._emit('change');
    }

    return this;
  },

  getBounds: function() {
    var centerXY = project(this._center.latitude, this._center.longitude, this._worldSize);

    var size = this.getSize();
    var halfWidth  = size.width/2;
    var halfHeight = size.height/2;

    var nw = unproject(centerXY.x - halfWidth, centerXY.y - halfHeight, this._worldSize);
    var se = unproject(centerXY.x + halfWidth, centerXY.y + halfHeight, this._worldSize);

    return {
      n: nw.latitude,
      w: nw.longitude,
      s: se.latitude,
      e: se.longitude
    };
  },

  setSize: function(size) {
    var canvas = gl.canvas;
    if (size.width !== this._size.width || size.height !== this._size.height) {
      canvas.width  = this._size.width  = size.width;
      canvas.height = this._size.height = size.height;
      gl.viewport(0, 0, size.width, size.height);
      this._projection = Matrix.perspective(20, size.width, size.height, 40000);
      this._emit('resize');
    }

    return this;
  },

  getSize: function() {
    return this._size;
  },

  getOrigin: function() {
    return this._origin;
  },

  getRotation: function() {
    return this._rotation;
  },

  setRotation: function(rotation) {
    rotation = parseFloat(rotation)%360;
    if (this._rotation !== rotation) {
      this._rotation = rotation;
      this._emit('change');
    }
    return this;
  },

  getTilt: function() {
    return this._tilt;
  },

  setTilt: function(tilt) {
    tilt = clamp(parseFloat(tilt), 0, 70);
    if (this._tilt !== tilt) {
      this._tilt = tilt;
      this._emit('change');
    }
    return this;
  },

  getContext: function() {
    return gl;
  },

  destroy: function() {
    var canvas = gl.canvas;
    canvas.parentNode.removeChild(canvas);
    gl = null;

    // TODO: stop render loop
//  clearInterval(...);
    this._listeners = null;

    for (var i = 0; i < this._layers.length; i++) {
      this._layers[i].destroy();
    }
    this._layers = null;
  }
};
